Enable `if` and `match` in constants behind a feature flag

ecstatic-morse opened the pull request:
This PR is an initial implementation of #49146. It introduces a `const_if_match` feature flag and does the following if it is enabled:
- Allows `Downcast` projections, `SwitchInt` terminators and `FakeRead`s for matched places through the MIR const-checker.
- Allows `if` and `match` expressions through the HIR const-checker.
- Stops converting `&&` to `&` and `||` to `|` in `const` and `static` items.

As a result, the following operations are now allowed in a const context behind the feature flag:
- `if` and `match`
- short circuiting logic operators (`&&` and `||`)
- the `assert` and `debug_assert` macros (if the `const_panic` feature flag is also enabled)

However, the following operations remain forbidden:
- `while`, `loop` and `for` (see #52000)
- the `?` operator (calls `From::from` on its error variant)
- the `assert_eq` and `assert_ne` macros, along with their `debug` variants (calls `fmt::Debug`)

This PR is possible now that we use dataflow for const qualification (see #64470 and #66385).

r? @oli-obk 
cc @rust-lang/wg-const-eval @eddyb

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=13804&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13804/logs/67), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/13804/logs/67)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-18T05:47:26.6871374Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-18T05:47:27.6538045Z ##[command]git config gc.auto 0
2019-11-18T05:47:27.6543728Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-18T05:47:27.6548617Z ##[command]git config --get-all http.proxy
2019-11-18T05:47:27.6553947Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66507/merge:refs/remotes/pull/66507/merge
---
2019-11-18T06:43:45.2183360Z .................................................................................................... 1600/9264
2019-11-18T06:43:51.0550431Z .................................................................................................... 1700/9264
2019-11-18T06:44:02.7897276Z .....................i.............................................................................. 1800/9264
2019-11-18T06:44:09.5391576Z .................................................................................................... 1900/9264
2019-11-18T06:44:24.0259743Z ......iiiii......................................................................................... 2000/9264
2019-11-18T06:44:33.1164149Z .................................................................................................... 2200/9264
2019-11-18T06:44:35.6999536Z .................................................................................................... 2300/9264
2019-11-18T06:44:41.4417312Z .................................................................................................... 2400/9264
2019-11-18T06:45:02.1587562Z .................................................................................................... 2500/9264
---
2019-11-18T06:47:40.8641421Z ......i...............i............................................................................. 4800/9264
2019-11-18T06:47:50.5436251Z .................................................................................................... 4900/9264
2019-11-18T06:47:55.7457985Z .................................................................................................... 5000/9264
2019-11-18T06:48:05.1570413Z .................................................................................................... 5100/9264
2019-11-18T06:48:10.6135206Z ..........ii.ii...........i......................................................................... 5200/9264
2019-11-18T06:48:20.0019737Z .................................................................................................... 5400/9264
2019-11-18T06:48:30.3299288Z ............................................................................................i....... 5500/9264
2019-11-18T06:48:38.2566171Z .................................................................................................... 5600/9264
2019-11-18T06:48:43.2933844Z .................................................................................................... 5700/9264
2019-11-18T06:48:43.2933844Z .................................................................................................... 5700/9264
2019-11-18T06:48:53.0268149Z ..............................................................................ii...i..ii...........i 5800/9264
2019-11-18T06:49:15.0265544Z .................................................................................................... 6000/9264
2019-11-18T06:49:22.8468137Z .................................................................................................... 6100/9264
2019-11-18T06:49:28.6764077Z .................................................................................................i.. 6200/9264
2019-11-18T06:49:41.6467919Z ii.................................................................................................. 6300/9264
---
2019-11-18T06:55:06.7888974Z  finished in 5.541
2019-11-18T06:55:06.8065976Z Check compiletest suite=codegen mode=codegen (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-18T06:55:06.9790109Z 
2019-11-18T06:55:06.9790915Z running 156 tests
2019-11-18T06:55:09.8159756Z iiii....iii......iii..iiii...i.............................i..i..................i....i...........ii 100/156
2019-11-18T06:55:11.6503670Z .i.i..iiii..............i.........iii.i.........ii......
2019-11-18T06:55:11.6506302Z 
2019-11-18T06:55:11.6507210Z  finished in 4.843
2019-11-18T06:55:11.6684502Z Check compiletest suite=codegen-units mode=codegen-units (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-18T06:55:11.8243847Z 
---
2019-11-18T06:55:13.7298317Z  finished in 2.061
2019-11-18T06:55:13.7469987Z Check compiletest suite=assembly mode=assembly (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-18T06:55:13.9014254Z 
2019-11-18T06:55:13.9014497Z running 9 tests
2019-11-18T06:55:13.9015222Z iiiiiiiii
2019-11-18T06:55:13.9015578Z 
2019-11-18T06:55:13.9015622Z  finished in 0.153
2019-11-18T06:55:13.9208308Z Check compiletest suite=incremental mode=incremental (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
2019-11-18T06:55:14.0943048Z 
2019-11-18T06:55:14.0943048Z 
2019-11-18T06:55:14.0943276Z running 109 tests
2019-11-18T06:55:31.2939135Z .................................................................................................... 100/109
2019-11-18T06:55:32.5675452Z .......F.
2019-11-18T06:55:32.5677175Z failures:
2019-11-18T06:55:32.5678194Z 
2019-11-18T06:55:32.5679767Z ---- [incremental] incremental/warnings-reemitted.rs stdout ----
2019-11-18T06:55:32.5681153Z thread '[incremental] incremental/warnings-reemitted.rs' panicked at 'Error annotations must look like `//[X]~` in incremental tests', src/libcore/option.rs:1185:5
2019-11-18T06:55:32.5682614Z 
2019-11-18T06:55:32.5682887Z 
2019-11-18T06:55:32.5683674Z failures:
2019-11-18T06:55:32.5685025Z     [incremental] incremental/warnings-reemitted.rs
2019-11-18T06:55:32.5685025Z     [incremental] incremental/warnings-reemitted.rs
2019-11-18T06:55:32.5685995Z 
2019-11-18T06:55:32.5687433Z test result: FAILED. 108 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
2019-11-18T06:55:32.5688546Z 
2019-11-18T06:55:32.5693876Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-18T06:55:32.5697807Z 
2019-11-18T06:55:32.5698890Z 
2019-11-18T06:55:32.5701323Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/incremental" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/incremental" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "incremental" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-18T06:55:32.5701620Z 
2019-11-18T06:55:32.5701649Z 
2019-11-18T06:55:32.5710414Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-18T06:55:32.5710505Z Build completed unsuccessfully in 1:01:56
2019-11-18T06:55:32.5710505Z Build completed unsuccessfully in 1:01:56
2019-11-18T06:55:32.5765259Z == clock drift check ==
2019-11-18T06:55:32.5786155Z   local time: Mon Nov 18 06:55:32 UTC 2019
2019-11-18T06:55:33.1003373Z   network time: Mon, 18 Nov 2019 06:55:33 GMT
2019-11-18T06:55:33.1004237Z == end clock drift check ==
2019-11-18T06:55:38.1292835Z 
2019-11-18T06:55:38.1386704Z ##[error]Bash exited with code '1'.
2019-11-18T06:55:38.1423978Z ##[section]Starting: Checkout
2019-11-18T06:55:38.1425643Z ==============================================================================
2019-11-18T06:55:38.1425715Z Task         : Get sources
2019-11-18T06:55:38.1425800Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Centril commented on src/librustc_mir/transform/qualify_min_const_fn.rs:
```diff
@@ -324,10 +326,17 @@ fn check_terminator(
+        | TerminatorKind::SwitchInt { .. }
+        => Ok(()),
+
         | TerminatorKind::Abort | TerminatorKind::Unreachable => {
```
@oli-obk Also if you remember, what was the reason for disallowing `::Abort | ::Unreachable`?

oli-obk replied:
probably because we have no const fn code actually hitting it and thus there being no way to test it

Centril replied:
Can't we use `-Z panic=abort`?

oli-obk replied:
that just removes the landing pads, but still goes through the regular panic invocation I think. It just doesn't unwind in the end.

matthewjasper replied:
Abort terminators are generated on `#[unwind(aborts)]` functions when compiling with `-C panic=unwind`, which I believe can happen for `const fn`.

oli-obk replied:
I opened https://github.com/rust-lang/rust/issues/66557

Centril commented on src/test/ui/consts/const-match-pattern-arm.stderr:
```diff
@@ -7,6 +7,9 @@ LL | |     Some(value) => true,
 LL | |     _ => false
 LL | | };
    | |_^
+   |
```
Can we gather all control flow tests into `consts/control-flow` so that it's better organized?

ecstatic-morse replied:
I've done this for the obvious files, and will try to do this from now on.

Centril commented on src/test/ui/consts/control-flow/basics.rs:
```diff
@@ -0,0 +1,43 @@
+// Test basic functionality of `if` and `match` in a const context.
+
+// run-pass
```
More test ideas (for other files -- and in other follow up PRs if necessary):

- Test all the various https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/enum.PatKind.html forms and some random walk of that. (No reason to believe it's different, but sometimes that's wrong; well except `Box` patterns).

- Also test `if let` (good idea to not assume a HIR->MIR desugar since that might change, and probably will...).

ecstatic-morse replied:
@jyn514, this would good be a good place to contribute if you're still motivated. "Interesting" cases could be added to `basics.rs` (or another file) that use more types of patterns in `match` and `if let`. If you push to a branch I can cherry pick them into this PR (keeping you as the author of course), or you can open a separate PR once this is merged.

Centril commented on src/test/ui/consts/control-flow/drop.rs:
```diff
@@ -0,0 +1,43 @@
+
+#![feature(const_if_match)]
+
+// OK: `x` is always moved into the final value and is not dropped inside the initializer.
```
I would prefer if we could separate the good/bad parts of the test into different files (general note).

ecstatic-morse replied:
It's now done for this test and I'll do this in the future.

Centril commented on src/test/ui/consts/control-flow/drop.rs:
```diff
@@ -0,0 +1,43 @@
+        None => x,
+    }
+};
+
```
More interesting cases:
```rust
match () {
    () => x,
}
```
and:
```rust
match x {
    Ok(x) | Err(x) => Ok(x),
}
```

Centril commented on src/test/ui/consts/control-flow/feature-gate-const-if-match.rs:
```diff
@@ -0,0 +1,27 @@
+    match 0 { 1 => 2, _ => 0 } //[stock]~ ERROR `match` is not allowed in a `const fn`
+}
+
+#[rustc_error]
```
Would prefer splitting the file and having one be `// check-pass` so that stabilization becomes easier later on.

ecstatic-morse replied:
I prefer the `// revisions:` approach to having two copies of the file while the feature flag exists.

Centril replied:
@ecstatic-morse In that case, please leave a note so that the file isn't removed whole-sale when we want to stabilize and so that it is converted into `// check-pass`.

ecstatic-morse replied:
There's a comment at the top of the file now.

Centril commented on src/test/ui/consts/control-flow/interior-mutability.rs:
```diff
@@ -0,0 +1,18 @@
+        x = Some(Cell::new(4));
+    }
+    x
+};
```
Would be good to repeat this for `if let` and `match` (again to not presume the MIR lowering).

Centril replied:
Did you miss this one?

ecstatic-morse replied:
If I marked it resolved, it was by accident. I don't think it adds anything to repeat each qualifs test three times, however.

oli-obk replied:
I agree. If the lowerings start differing in ways that are problematic here we're gonna see other problems, too.

Centril commented on src/test/ui/consts/control-flow/short-circuit-let.rs:
```diff
@@ -0,0 +1,18 @@
+    let mut x = 0;
+    let _  = true && { x = 1; false };
+    x
+};
```
Let's also test the values at compile time & run-time.

Would be also good to have e.g.
```rust
const fn foo() -> bool {
    let x = true && return false; // and similar for `||`
    x
}
```

Centril commented on src/test/ui/consts/miri_unleashed/enum_discriminants.rs:
```diff
@@ -3,6 +3,7 @@
 
 // miri unleashed warnings are not useful and change frequently, so they are silenced above.
```
What happens if you remove -Zunleash with the new feature enabled?

ecstatic-morse replied:
Good point. This should no longer require the flag.

Centril commented on src/test/ui/consts/single_variant_match_ice.stderr:
```diff
@@ -6,6 +6,9 @@ LL |   const FOO: u32 = match Foo::Prob {
    | |_^
+   |
+   = note: branching and conditionals are experimental in a const context
+   = help: add `#![feature(const_if_match)]` to the crate attributes to enable
```
Seems like this can be `// check-pass` now?

Centril commented on src/test/ui/issues/issue-46843.stderr:
```diff
@@ -8,6 +8,9 @@ LL | |     Thing::This => 1,
    | |_^
+   |
+   = note: branching and conditionals are experimental in a const context
+   = help: add `#![feature(const_if_match)]` to the crate attributes to enable
```
Let's enable the gate here (conditionally) and move the test into `control-flow/` ?

Centril commented on src/test/ui/issues/issue-50577.stderr:
```diff
@@ -4,6 +4,8 @@ error[E0744]: `match` is not allowed in a `const`
    |                ^^^^^^^^^^^^^^^^
    |
+   = note: branching and conditionals are experimental in a const context
+   = help: add `#![feature(const_if_match)]` to the crate attributes to enable
```
Would be good to conditionally switch on the gate here also.

Centril commented on src/tools/compiletest/src/errors.rs:
```diff
@@ -7,7 +7,9 @@ use std::io::BufReader;
 use std::path::Path;
```
Would prefer if we did the changes to compiletest in a different PR including with tests independent of the added feature gate to ensure compiletest is working properly.

ecstatic-morse replied:
See #66524

Centril commented:
Added the feature gate label. @ecstatic-morse Could you please mark relevant past PRs that lead up to this one with that label as well? (It's very helpful for tracking implementation history for stabilization report purposes.)

oli-obk commented on src/test/ui/consts/control-flow/interior-mutability.rs:
```diff
@@ -0,0 +1,18 @@
+
+use std::cell::Cell;
+
+// Ensure that *any* assignment to the return place of a value with interior mutability
```
Do we have a test for

```
const X: Option<Cell<i32>> = {
    let mut x = Some(Cell::new(4));
    if true {
        x = None;
    }
    x
};
```

?

ecstatic-morse replied:
Added [here](https://github.com/rust-lang/rust/pull/66507/files#diff-e2a59c2bcf3c9ce2dd92a137f9999680R16-R22).

oli-obk commented on src/librustc_passes/check_const.rs:
```diff
@@ -75,31 +115,62 @@ pub(crate) fn provide(providers: &mut Providers<'_>) {
+
+        match expr {
+            If | IfLet | Match => {
+                err.note("branching and conditionals are experimental in a const context");
```
there's a specific macro for generating feature gate errors. I think this should be used here instead of rerolling the logic

Centril replied:
See https://github.com/rust-lang/rust/pull/66507#discussion_r347256214 .

oli-obk commented on src/test/ui/consts/control-flow/drop.rs:
```diff
@@ -0,0 +1,43 @@
+    }
+};
+
+// BUG?: even though we move the `Vec` out of `x`, we don't mark `x` as no longer `NeedsDrop`.
```
do you want to open an issue about it or are you addressing it already?

ecstatic-morse replied:
We reject this for the same reason we reject:

```rust
const fn foo() -> Vec<i32> {
    let vec_and_bool = (Vec::new(), true);
    vec_and_bool.0
}
```

My thought was that we could special case that a move out of the only field of a struct or enum variant also clears `Drop` for the container. We may want to defer this to a follow-up PR? I'm not quite sure what all the implications of this are.

oli-obk replied:
Oh yea, please defer. Being overly restrictive at the start is totally the right way

ecstatic-morse replied:
I added the reduced version as a failing test case.

oli-obk commented on src/librustc_mir/transform/qualify_min_const_fn.rs:
```diff
@@ -216,7 +216,9 @@ fn check_statement(
+        => {
             Err((span, "loops and conditional expressions are not stable in const fn".into()))
         }
 
```
yea we should probably match on these exhaustively

maybe a single arm for the `FakeRead` and then take its fields apart

ecstatic-morse replied:
I think you two are discussing different things here? It seems like we should indeed check the LHS of `SetDiscriminant`, although we need to do this in `check_consts` as well, and I think it's unrelated to this PR. We should be using a `Visitor` here and just calling `super_statement`, but this is blocked on #65572.

`FakeRead` is handled properly at the moment IMO.

oli-obk replied:
Yea I didn't read the patterns correctly. Ignore that comment

oli-obk replied:
Opened https://github.com/rust-lang/rust/issues/66556

ecstatic-morse commented on src/librustc_passes/check_const.rs:
```diff
@@ -75,31 +115,62 @@ pub(crate) fn provide(providers: &mut Providers<'_>) {
+        );
+
+        match expr {
+            If | IfLet | Match => {
```
My bad. I didn't know that existed.

jyn514 commented:
@ecstatic-morse this is really exciting! I still want to contribute tests but my laptop is currently broken so I'll be helping only sporadically.

Where would I start? I see you added tests in `src/test/ui/consts/control-flow/basics.rs` so I can follow that format, but what sort of things need to be tested? I would mostly be using this for enum matching (things like https://github.com/bytecodealliance/target-lexicon/blob/master/src/targets.rs#L126) but I imagine that isn't the hard part of this feature.

ecstatic-morse commented on src/librustc_passes/check_const.rs:
```diff
@@ -11,16 +11,56 @@ use rustc::hir::def_id::DefId;
 
+/// An expression that is not *always* legal in a const context.
+#[derive(Clone, Copy)]
+enum NonConstExpr {
```
This is mutually exclusive with the change below (`while let` => `while`) unless we want to change `LoopSource::name`, which is probably fine.

Centril replied:
I would prefer for this one to take priority and perhaps think about the output below later ; my main point below was just that we should be consistent wrt if and while :)

ecstatic-morse commented:
I've rebased with @Centril's changes to the HIR const-checker applied. I now use `emit_feature_err` instead of custom logic. I also moved the `compiletest` changes into their own PR (#66524). I squashed all the test changes into one commit since the error messages are in flux and rebasing after each `bless` became a pain.

oli-obk commented:
Does this PR still count as WIP? Seems ready to me

ecstatic-morse commented:
@oli-obk Yeah, I think this is ready to go. I'd like to have a few more tests for interior mutability/`Drop`, but it's quite hard to come up with novel ones. Hopefully nightly users are more creative :smile:.

ecstatic-morse commented:
Oh, this will have to be merged after #66524. They shouldn't conflict though.

Centril commented:
@ecstatic-morse would be good to address https://github.com/rust-lang/rust/pull/66507#discussion_r348463724 first tho.

goto-bus-stop commented on src/doc/unstable-book/src/language-features/const-if-match.md:
```diff
@@ -0,0 +1,14 @@
+
+The tracking issue for this feature is: [#49146]
+
+[#49146]: https://github.com/rust-lang/rust/issues/49147
```
```suggestion
[#49146]: https://github.com/rust-lang/rust/issues/49146
```

bors commented:
:umbrella: The latest upstream changes (presumably #66578) made this pull request unmergeable. Please resolve the merge conflicts.

Centril commented on src/libsyntax/feature_gate/active.rs:
```diff
@@ -529,6 +529,9 @@ declare_features! (
     (active, register_tool, "1.41.0", Some(66079), None),
 
+    /// Allows the use of `if` and `match` in constants.
+    (active, const_if_match, "1.41.0", Some(49146), None),
```
Btw: a useful trick for the future, if you want to avoid rebasing due to this, is to move this before `never_type_fallback`.

rust-highfive commented:
The job `x86_64-gnu-llvm-6.0` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=14110&view=logs&jobId=d0d954b5-f111-5dc4-4d76-03b6c9d0cf7e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14110/logs/66), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/14110/logs/66)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2019-11-21T23:24:47.3008807Z ##[command]git remote add origin https://github.com/rust-lang/rust
2019-11-21T23:24:48.0725402Z ##[command]git config gc.auto 0
2019-11-21T23:24:48.0729585Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2019-11-21T23:24:48.0734229Z ##[command]git config --get-all http.proxy
2019-11-21T23:24:48.0739881Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/66507/merge:refs/remotes/pull/66507/merge
---
2019-11-22T00:25:30.1450355Z .................................................................................................... 1600/9278
2019-11-22T00:25:35.0259555Z .................................................................................................... 1700/9278
2019-11-22T00:25:48.1920285Z ...........................i........................................................................ 1800/9278
2019-11-22T00:25:54.9704915Z .................................................................................................... 1900/9278
2019-11-22T00:26:09.4140055Z ............iiiii................................................................................... 2000/9278
2019-11-22T00:26:19.0124743Z .................................................................................................... 2200/9278
2019-11-22T00:26:21.5187022Z .................................................................................................... 2300/9278
2019-11-22T00:26:26.6422144Z .................................................................................................... 2400/9278
2019-11-22T00:26:48.2036519Z .................................................................................................... 2500/9278
---
2019-11-22T00:29:28.8651495Z ............i...............i....................................................................... 4800/9278
2019-11-22T00:29:39.6432697Z .................................................................................................... 4900/9278
2019-11-22T00:29:45.1443444Z .................................................................................................... 5000/9278
2019-11-22T00:29:54.7092063Z .................................................................................................... 5100/9278
2019-11-22T00:30:00.8071050Z .................ii.ii...........i.................................................................. 5200/9278
2019-11-22T00:30:10.0652922Z .................................................................................................... 5400/9278
2019-11-22T00:30:20.9805431Z ...............F...................................................................................i 5500/9278
2019-11-22T00:30:28.9067102Z .................................................................................................... 5600/9278
2019-11-22T00:30:34.7351969Z .................................................................................................... 5700/9278
2019-11-22T00:30:34.7351969Z .................................................................................................... 5700/9278
2019-11-22T00:30:45.2815902Z .....................................................................................ii...i..ii..... 5800/9278
2019-11-22T00:31:08.1491749Z .................................................................................................... 6000/9278
2019-11-22T00:31:16.2418734Z .................................................................................................... 6100/9278
2019-11-22T00:31:24.7287135Z .................................................................................................... 6200/9278
2019-11-22T00:31:24.7287135Z .................................................................................................... 6200/9278
2019-11-22T00:31:40.4345569Z ........i..ii....................................................................................... 6300/9278
2019-11-22T00:31:59.7907216Z ............................................................................i....................... 6500/9278
2019-11-22T00:32:02.2219295Z .................................................................................................... 6600/9278
2019-11-22T00:32:04.5895657Z .................................................................i.................................. 6700/9278
2019-11-22T00:32:07.6714622Z .................................................................................................... 6800/9278
---
2019-11-22T00:36:56.7331437Z diff of stderr:
2019-11-22T00:36:56.7331474Z 
2019-11-22T00:36:56.7331522Z 29    |             ^^^^^
2019-11-22T00:36:56.7331567Z 30 
2019-11-22T00:36:56.7331671Z 31 error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7332906Z +   --> $DIR/loop-break-value.rs:36:12
2019-11-22T00:36:56.7332988Z 33    |
2019-11-22T00:36:56.7333035Z 34 LL |         if break () {
2019-11-22T00:36:56.7333035Z 34 LL |         if break () {
2019-11-22T00:36:56.7333093Z 35    |            ^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7333199Z 40    |            ^^^^^
2019-11-22T00:36:56.7333243Z 41 
2019-11-22T00:36:56.7333243Z 41 
2019-11-22T00:36:56.7333293Z 42 error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7333825Z +   --> $DIR/loop-break-value.rs:41:9
2019-11-22T00:36:56.7333874Z 44    |
2019-11-22T00:36:56.7333936Z 45 LL |         break None;
2019-11-22T00:36:56.7333936Z 45 LL |         break None;
2019-11-22T00:36:56.7333993Z 46    |         ^^^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7334073Z 51    |         ^^^^^
2019-11-22T00:36:56.7334134Z 52 
2019-11-22T00:36:56.7334134Z 52 
2019-11-22T00:36:56.7334182Z 53 error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7334728Z +   --> $DIR/loop-break-value.rs:47:13
2019-11-22T00:36:56.7334777Z 55    |
2019-11-22T00:36:56.7334777Z 55    |
2019-11-22T00:36:56.7335034Z 56 LL |             break 'while_let_loop "nope";
2019-11-22T00:36:56.7335113Z 57    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7335182Z 
2019-11-22T00:36:56.7335231Z The actual stderr differed from the expected stderr.
2019-11-22T00:36:56.7335594Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/loops/loop-break-value/loop-break-value.stderr
2019-11-22T00:36:56.7335594Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/loops/loop-break-value/loop-break-value.stderr
2019-11-22T00:36:56.7335884Z To update references, rerun the tests and pass the `--bless` flag
2019-11-22T00:36:56.7336188Z To only update this specific test, also pass `--test-args loops/loop-break-value.rs`
2019-11-22T00:36:56.7336294Z error: 1 errors occurred comparing output.
2019-11-22T00:36:56.7336344Z status: exit code: 1
2019-11-22T00:36:56.7336344Z status: exit code: 1
2019-11-22T00:36:56.7337570Z command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/loops/loop-break-value.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/loops/loop-break-value" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/loops/loop-break-value/auxiliary" "-A" "unused"
2019-11-22T00:36:56.7337982Z ------------------------------------------
2019-11-22T00:36:56.7338023Z 
2019-11-22T00:36:56.7338275Z ------------------------------------------
2019-11-22T00:36:56.7338324Z stderr:
2019-11-22T00:36:56.7338324Z stderr:
2019-11-22T00:36:56.7338587Z ------------------------------------------
2019-11-22T00:36:56.7338640Z warning: denote infinite loops with `loop { ... }`
2019-11-22T00:36:56.7338998Z    |
2019-11-22T00:36:56.7338998Z    |
2019-11-22T00:36:56.7339283Z LL |     'while_loop: while true { //~ WARN denote infinite loops with
2019-11-22T00:36:56.7339340Z    |     ^^^^^^^^^^^^^^^^^^^^^^^ help: use `loop`
2019-11-22T00:36:56.7339403Z    |
2019-11-22T00:36:56.7339452Z    = note: `#[warn(while_true)]` on by default
2019-11-22T00:36:56.7339485Z 
2019-11-22T00:36:56.7339534Z error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7339875Z    |
2019-11-22T00:36:56.7339875Z    |
2019-11-22T00:36:56.7339927Z LL |         break (); //~ ERROR `break` with value from a `while` loop
2019-11-22T00:36:56.7340001Z    |         ^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7340053Z    |
2019-11-22T00:36:56.7340109Z help: instead, use `break` on its own without a value inside this `while` loop
2019-11-22T00:36:56.7340175Z    |
2019-11-22T00:36:56.7340245Z LL |         break; //~ ERROR `break` with value from a `while` loop
2019-11-22T00:36:56.7340341Z 
2019-11-22T00:36:56.7340341Z 
2019-11-22T00:36:56.7340390Z error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7340728Z    |
2019-11-22T00:36:56.7340989Z LL |             break 'while_loop 123;
2019-11-22T00:36:56.7340989Z LL |             break 'while_loop 123;
2019-11-22T00:36:56.7341048Z    |             ^^^^^^^^^^^^^^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7341099Z    |
2019-11-22T00:36:56.7341167Z help: instead, use `break` on its own without a value inside this `while` loop
2019-11-22T00:36:56.7341262Z LL |             break;
2019-11-22T00:36:56.7341325Z    |             ^^^^^
2019-11-22T00:36:56.7341355Z 
2019-11-22T00:36:56.7341355Z 
2019-11-22T00:36:56.7341404Z error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7341760Z    |
2019-11-22T00:36:56.7341760Z    |
2019-11-22T00:36:56.7341813Z LL |         if break () { //~ ERROR `break` with value from a `while` loop
2019-11-22T00:36:56.7342155Z    |            ^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7342231Z    |
2019-11-22T00:36:56.7342284Z help: instead, use `break` on its own without a value inside this `while` loop
2019-11-22T00:36:56.7342332Z    |
2019-11-22T00:36:56.7342399Z LL |         if break { //~ ERROR `break` with value from a `while` loop
2019-11-22T00:36:56.7342479Z 
2019-11-22T00:36:56.7342479Z 
2019-11-22T00:36:56.7342528Z error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7342929Z    |
2019-11-22T00:36:56.7342975Z LL |         break None;
2019-11-22T00:36:56.7342975Z LL |         break None;
2019-11-22T00:36:56.7343044Z    |         ^^^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7343225Z    |
2019-11-22T00:36:56.7343336Z help: instead, use `break` on its own without a value inside this `while` loop
2019-11-22T00:36:56.7343452Z LL |         break;
2019-11-22T00:36:56.7343499Z    |         ^^^^^
2019-11-22T00:36:56.7343529Z 
2019-11-22T00:36:56.7343529Z 
2019-11-22T00:36:56.7343593Z error[E0571]: `break` with value from a `while` loop
2019-11-22T00:36:56.7343957Z    |
2019-11-22T00:36:56.7343957Z    |
2019-11-22T00:36:56.7344227Z LL |             break 'while_let_loop "nope";
2019-11-22T00:36:56.7344290Z    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7344344Z    |
2019-11-22T00:36:56.7344420Z help: instead, use `break` on its own without a value inside this `while` loop
2019-11-22T00:36:56.7344514Z LL |             break;
2019-11-22T00:36:56.7344559Z    |             ^^^^^
2019-11-22T00:36:56.7344615Z 
2019-11-22T00:36:56.7344615Z 
2019-11-22T00:36:56.7344671Z error[E0571]: `break` with value from a `for` loop
2019-11-22T00:36:56.7345019Z    |
2019-11-22T00:36:56.7345019Z    |
2019-11-22T00:36:56.7345071Z LL |         break (); //~ ERROR `break` with value from a `for` loop
2019-11-22T00:36:56.7345130Z    |         ^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7345193Z    |
2019-11-22T00:36:56.7345246Z help: instead, use `break` on its own without a value inside this `for` loop
2019-11-22T00:36:56.7345295Z    |
2019-11-22T00:36:56.7345403Z LL |         break; //~ ERROR `break` with value from a `for` loop
2019-11-22T00:36:56.7345500Z 
2019-11-22T00:36:56.7345500Z 
2019-11-22T00:36:56.7345548Z error[E0571]: `break` with value from a `for` loop
2019-11-22T00:36:56.7345892Z    |
2019-11-22T00:36:56.7345892Z    |
2019-11-22T00:36:56.7345937Z LL |         break [()];
2019-11-22T00:36:56.7346008Z    |         ^^^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7346072Z    |
2019-11-22T00:36:56.7346125Z help: instead, use `break` on its own without a value inside this `for` loop
2019-11-22T00:36:56.7346234Z LL |         break;
2019-11-22T00:36:56.7346280Z    |         ^^^^^
2019-11-22T00:36:56.7346309Z 
2019-11-22T00:36:56.7346309Z 
2019-11-22T00:36:56.7346357Z error[E0571]: `break` with value from a `for` loop
2019-11-22T00:36:56.7346704Z    |
2019-11-22T00:36:56.7346952Z LL |             break 'for_loop Some(17);
2019-11-22T00:36:56.7346952Z LL |             break 'for_loop Some(17);
2019-11-22T00:36:56.7347028Z    |             ^^^^^^^^^^^^^^^^^^^^^^^^ can only break with a value inside `loop` or breakable block
2019-11-22T00:36:56.7347079Z    |
2019-11-22T00:36:56.7347130Z help: instead, use `break` on its own without a value inside this `for` loop
2019-11-22T00:36:56.7347241Z LL |             break;
2019-11-22T00:36:56.7347302Z    |             ^^^^^
2019-11-22T00:36:56.7347332Z 
2019-11-22T00:36:56.7347393Z error[E0308]: mismatched types
2019-11-22T00:36:56.7347393Z error[E0308]: mismatched types
2019-11-22T00:36:56.7347665Z   --> /checkout/src/test/ui/loops/loop-break-value.rs:2:31
2019-11-22T00:36:56.7347715Z    |
2019-11-22T00:36:56.7347778Z LL |     let val: ! = loop { break break; };
2019-11-22T00:36:56.7347833Z    |                               ^^^^^ expected `!`, found `()`
2019-11-22T00:36:56.7347940Z    = note:   expected type `!`
2019-11-22T00:36:56.7347989Z            found unit type `()`
2019-11-22T00:36:56.7348020Z 
2019-11-22T00:36:56.7348064Z error[E0308]: mismatched types
---
2019-11-22T00:36:56.7348561Z 
2019-11-22T00:36:56.7348700Z error[E0308]: mismatched types
2019-11-22T00:36:56.7349079Z   --> /checkout/src/test/ui/loops/loop-break-value.rs:14:15
2019-11-22T00:36:56.7349157Z    |
2019-11-22T00:36:56.7349206Z LL |         break "asdf"; //~ ERROR mismatched types
2019-11-22T00:36:56.7349259Z    |               ^^^^^^ expected `i32`, found `&str`
2019-11-22T00:36:56.7349355Z error[E0308]: mismatched types
2019-11-22T00:36:56.7349658Z   --> /checkout/src/test/ui/loops/loop-break-value.rs:19:31
2019-11-22T00:36:56.7349709Z    |
2019-11-22T00:36:56.7349709Z    |
2019-11-22T00:36:56.7350003Z LL |             break 'outer_loop "nope"; //~ ERROR mismatched types
2019-11-22T00:36:56.7350062Z    |                               ^^^^^^ expected `i32`, found `&str`
2019-11-22T00:36:56.7350143Z error[E0308]: mismatched types
2019-11-22T00:36:56.7350431Z   --> /checkout/src/test/ui/loops/loop-break-value.rs:71:26
2019-11-22T00:36:56.7350481Z    |
2019-11-22T00:36:56.7350481Z    |
2019-11-22T00:36:56.7350749Z LL |                 break 'c 123; //~ ERROR mismatched types
2019-11-22T00:36:56.7350841Z    |                          ^^^ expected `()`, found integer
2019-11-22T00:36:56.7350920Z error[E0308]: mismatched types
2019-11-22T00:36:56.7351206Z   --> /checkout/src/test/ui/loops/loop-break-value.rs:78:15
2019-11-22T00:36:56.7351256Z    |
2019-11-22T00:36:56.7351256Z    |
2019-11-22T00:36:56.7351306Z LL |         break (break, break); //~ ERROR mismatched types
2019-11-22T00:36:56.7351360Z    |               ^^^^^^^^^^^^^^ expected `()`, found tuple
2019-11-22T00:36:56.7351468Z    = note: expected unit type `()`
2019-11-22T00:36:56.7351468Z    = note: expected unit type `()`
2019-11-22T00:36:56.7351518Z                   found tuple `(!, !)`
2019-11-22T00:36:56.7351613Z error[E0308]: mismatched types
2019-11-22T00:36:56.7352296Z   --> /checkout/src/test/ui/loops/loop-break-value.rs:83:15
2019-11-22T00:36:56.7352359Z    |
2019-11-22T00:36:56.7352427Z LL |         break 2; //~ ERROR mismatched types
---
2019-11-22T00:36:56.7353297Z    |
2019-11-22T00:36:56.7353346Z LL |         break; //~ ERROR mismatched types
2019-11-22T00:36:56.7353412Z    |         ^^^^^
2019-11-22T00:36:56.7353457Z    |         |
2019-11-22T00:36:56.7353506Z    |         expected integer, found `()`
2019-11-22T00:36:56.7353575Z    |         help: give it a value of the expected type: `break value`
2019-11-22T00:36:56.7353658Z error: aborting due to 16 previous errors
2019-11-22T00:36:56.7353690Z 
2019-11-22T00:36:56.7353756Z Some errors have detailed explanations: E0308, E0571.
2019-11-22T00:36:56.7354288Z For more information about an error, try `rustc --explain E0308`.
---
2019-11-22T00:36:56.7361281Z thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:537:22
2019-11-22T00:36:56.7361617Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
2019-11-22T00:36:56.7376513Z 
2019-11-22T00:36:56.7376782Z 
2019-11-22T00:36:56.7379312Z command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-6.0/bin/FileCheck" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "6.0.0\n" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"
2019-11-22T00:36:56.7380087Z 
2019-11-22T00:36:56.7380246Z 
2019-11-22T00:36:56.7385411Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test
2019-11-22T00:36:56.7385647Z Build completed unsuccessfully in 1:05:48
2019-11-22T00:36:56.7385647Z Build completed unsuccessfully in 1:05:48
2019-11-22T00:36:56.7436453Z == clock drift check ==
2019-11-22T00:36:56.7456543Z   local time: Fri Nov 22 00:36:56 UTC 2019
2019-11-22T00:36:57.0280288Z   network time: Fri, 22 Nov 2019 00:36:57 GMT
2019-11-22T00:36:57.0285926Z == end clock drift check ==
2019-11-22T00:36:57.8599199Z 
2019-11-22T00:36:57.8730299Z ##[error]Bash exited with code '1'.
2019-11-22T00:36:57.8767554Z ##[section]Starting: Checkout
2019-11-22T00:36:57.8769322Z ==============================================================================
2019-11-22T00:36:57.8769374Z Task         : Get sources
2019-11-22T00:36:57.8769434Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

ecstatic-morse commented:
@oli-obk Can you do a final review of this when you have the time?

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit b09bb1569b23eaadcf22d7420f1ba9872c1088f7 has been approved by `oli-obk`

<!-- @bors r=oli-obk b09bb1569b23eaadcf22d7420f1ba9872c1088f7 -->
<!-- homu: {"type":"Approved","sha":"b09bb1569b23eaadcf22d7420f1ba9872c1088f7","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit b09bb1569b23eaadcf22d7420f1ba9872c1088f7 with merge 6d523ee501a1ae30428d22ce4fc3c1f720e896e6...
<!-- homu: {"type":"BuildStarted","head_sha":"b09bb1569b23eaadcf22d7420f1ba9872c1088f7","merge_sha":"6d523ee501a1ae30428d22ce4fc3c1f720e896e6"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14198)
Approved by: oli-obk
Pushing 6d523ee501a1ae30428d22ce4fc3c1f720e896e6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=14198"},"merge_sha":"6d523ee501a1ae30428d22ce4fc3c1f720e896e6"} -->
